### PR TITLE
Add cpu-quota support

### PR DESCRIPTION
**Pull request by damianopezzotti**

### CODE REVIEW

**Comment by davidxia** [comment]
@damianopezzotti thanks! Can we keep the memory, memorySwap, etc as clients might use them?

**Comment by davidxia** [comment]
Can you write a test for cpuquota?

**Comment by damianopezzotti** [comment]
In the ContainerConfig class?


2015-11-17 19:12 GMT+01:00 David Xia <notifications@github.com>:

> @damianopezzotti <https://github.com/damianopezzotti> thanks! Can we keep
> the memory, memorySwap, etc as clients might use them?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/docker-client/pull/286#issuecomment-157456871>
> .
>



-- 
Damiano Pezzotti
...................................... . .
Software Architect | Alambitco
mobile +39 333 2116258
fax +39 02 700432209
email damy@alambitco.com
web www.alambitco.com
flickr www.flickr.com/photos/damianopezzotti
linkedin www.linkedin.com/in/damianopezzotti
pgp id 0x12DFD989

**Comment by davidxia** [comment]
yes

On Tue, Nov 17, 2015 at 3:00 PM, damianopezzotti <notifications@github.com>
wrote:

> In the ContainerConfig class?
>
>
> 2015-11-17 19:12 GMT+01:00 David Xia <notifications@github.com>:
>
> > @damianopezzotti <https://github.com/damianopezzotti> thanks! Can we
> keep
> > the memory, memorySwap, etc as clients might use them?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/spotify/docker-client/pull/286#issuecomment-157456871>
> > .
> >
>
>
>
> --
> Damiano Pezzotti
> ...................................... . .
> Software Architect | Alambitco
> mobile +39 333 2116258
> fax +39 02 700432209
> email damy@alambitco.com
> web www.alambitco.com
> flickr www.flickr.com/photos/damianopezzotti
> linkedin www.linkedin.com/in/damianopezzotti
> pgp id 0x12DFD989
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/docker-client/pull/286#issuecomment-157488693>
> .
>



-- 
David Xia
Software Engineer
dxia@spotify.com

**Comment by damianopezzotti** [comment]
Ok, I restore the attributes and the new one, but it seems that these
attributes are not used

2015-11-17 21:00 GMT+01:00 David Xia <notifications@github.com>:

> yes
>
> On Tue, Nov 17, 2015 at 3:00 PM, damianopezzotti <notifications@github.com
> >
> wrote:
>
> > In the ContainerConfig class?
> >
> >
> > 2015-11-17 19:12 GMT+01:00 David Xia <notifications@github.com>:
> >
> > > @damianopezzotti <https://github.com/damianopezzotti> thanks! Can we
> > keep
> > > the memory, memorySwap, etc as clients might use them?
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> > https://github.com/spotify/docker-client/pull/286#issuecomment-157456871
> >
> > > .
> > >
> >
> >
> >
> > --
> > Damiano Pezzotti
> > ...................................... . .
> > Software Architect | Alambitco
> > mobile +39 333 2116258
> > fax +39 02 700432209
> > email damy@alambitco.com
> > web www.alambitco.com
> > flickr www.flickr.com/photos/damianopezzotti
> > linkedin www.linkedin.com/in/damianopezzotti
> > pgp id 0x12DFD989
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/spotify/docker-client/pull/286#issuecomment-157488693>
> > .
> >
>
>
>
> --
> David Xia
> Software Engineer
> dxia@spotify.com
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/docker-client/pull/286#issuecomment-157488823>
> .
>



-- 
Damiano Pezzotti
...................................... . .
Software Architect | Alambitco
mobile +39 333 2116258
fax +39 02 700432209
email damy@alambitco.com
web www.alambitco.com
flickr www.flickr.com/photos/damianopezzotti
linkedin www.linkedin.com/in/damianopezzotti
pgp id 0x12DFD989

**Comment by damianopezzotti** [comment]
Test Added, properties restored

File: src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java [hunk 1]
```diff
@@ -1134,6 +1134,7 @@ public void testContainerWithHostConfig() throws Exception {
         .publishAllPorts(publishAllPorts)
         .dns(dns)
         .cpuShares((long) 4096)
+        .cpuQuota((long) 50000)
```
Comment (davidxia) [inline src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java]: For this test to be meaningful, we should inspect the container and see this value right? I don't know why we didn't do it for cpuShares above.

Comment (damianopezzotti) [inline src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java]: I added the assertion. On my local machine test passes, while on CircleCI I have a failure...

**Comment by mattnworb** [comment]
@damianopezzotti the test passes for me locally as well. Are you able to tell Circle to rebuild your commit? (or maybe just amend it and force-push to trick it).

Otherwise LGTM.

**Comment by davidxia** [comment]
I've seen circleCI's custom docker sometimes not support certain options. This might be happening here, but I haven't looked into it. If this is the case, you can add an `assume` or some docker api version checking in the test.

**Comment by damianopezzotti** [comment]
The test fails again on quota assertion.

From CircleCI logs I have the following information:

Client:
 Version:      1.8.2-circleci
 API version:  1.20

Server:
 Version:      1.8.2-circleci
 API version:  1.20

I can force the test to check api 1.21 but when CircleCI upgrade Docker to
version 1.9 we could have the same problem again...




2015-11-18 21:12 GMT+01:00 David Xia <notifications@github.com>:

> I've seen circleCI's custom docker sometimes not support certain options.
> This might be happening here, but I haven't looked into it. If this is the
> case, you can add an assume or some docker api version checking in the
> test.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/docker-client/pull/286#issuecomment-157847978>
> .
>



-- 
Damiano Pezzotti
...................................... . .
Software Architect | Alambitco
mobile +39 333 2116258
fax +39 02 700432209
email damy@alambitco.com
web www.alambitco.com
flickr www.flickr.com/photos/damianopezzotti
linkedin www.linkedin.com/in/damianopezzotti
pgp id 0x12DFD989

**Comment by mattnworb** [comment]
@damianopezzotti I'll confirm with CircleCI if they have disabled support for `cpuQuota`. If they indeed don't have support for it then you could just add an assumption in the test that it is not running in Circle like `testPullBadImage()` has.

**Comment by mattnworb** [comment]
@damianopezzotti also thank you very much for the patch :)

**Comment by damianopezzotti** [comment]
I excluded cpuQuota test for CircleCI

**Comment by mattnworb** [comment]
Thanks @damianopezzotti , I'd like to confirm for sure with Docker if this is the reason why the test fails before merging.

**Comment by mattnworb** [comment]
Received a response from Circle:

> We are running LXC unprivileged container to run your build and Docker is running on top of that. It's possible that the limitation comes from LXC, but not docker.

> Do you have code to show how cpu quota is set?

I sent back a link to this PR.

**Comment by mattnworb** [comment]
From Circle:

> 
Hi Matt,

> After taking a look at that it's definitely due to this running in an unprivileged LXC container and wouldn't be possible to set in CircleCI. Sorry about that!

Thanks again @damianopezzotti for the patch and also for taking care of all the necessary revisions

**Pull request merged**